File: lost-and-found/backend/app/admin_routes.py
from fastapi import APIRouter, Depends, HTTPException, status, Form, Response, Body
from sqlalchemy.orm import Session
from database import get_db
from models.user import User
from models.post import Post
from models.comment import Comment
from typing import List, Optional, Dict
import logging
from datetime import datetime, timedelta
import jwt
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
import os
from dotenv import load_dotenv
from pydantic import BaseModel, Field
from fastapi.middleware.cors import CORSMiddleware

# Load environment variables
load_dotenv()

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api")
security = HTTPBearer()

# JWT settings
SECRET_KEY = os.getenv("JWT_SECRET_KEY", "your-secret-key")
ALGORITHM = "HS256"

# Request models
class AdminLoginRequest(BaseModel):
    username: str
    password: str

def verify_admin_token(credentials: HTTPAuthorizationCredentials = Depends(security)):
    try:
        token = credentials.credentials
        payload = jwt.decode(token, SECRET_KEY, algorithms=[ALGORITHM])
        is_admin = payload.get("is_admin", False)
        if not is_admin:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Not authorized as admin"
            )
        return payload
    except jwt.JWTError:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Invalid authentication token"
        )

@router.get("/admin/setup-admin")
def setup_admin(db: Session = Depends(get_db)):
    """Create initial admin user if none exists"""
    logger.info("Setting up admin user...")
    
    # Check if admin already exists
    admin = db.query(User).filter(User.is_admin == True).first()
    if admin:
        logger.warning("Admin user already exists")
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Admin already exists"
        )
    
    # Create admin user
    admin = User(
        username="admin",
        password="umbc_admin_2024",
        email="admin@umbc.edu",
        is_admin=True
    )
    db.add(admin)
    db.commit()
    db.refresh(admin)
    logger.info("Admin user created successfully")
    return {"message": "Admin user created successfully"}

@router.post("/admin/login")
async def admin_login(request: AdminLoginRequest, db: Session = Depends(get_db)):
    """Handle admin login and return JWT token"""
    logger.info(f"Admin login attempt for username: {request.username}")
    
    # Find admin user
    admin = db.query(User).filter(
        User.username == request.username,
        User.is_admin == True
    ).first()
    
    if not admin or admin.password != request.password:
        logger.warning(f"Invalid admin login attempt for username: {request.username}")
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Invalid credentials"
        )
    
    # Generate token
    token_data = {
        "sub": admin.username,
        "is_admin": True,
        "exp": datetime.utcnow() + timedelta(days=1)
    }
    token = jwt.encode(token_data, SECRET_KEY, algorithm=ALGORITHM)
    logger.info(f"Admin login successful for username: {request.username}")
    
    return {"token": token}

@router.get("/admin/posts")
def get_posts(db: Session = Depends(get_db), _: dict = Depends(verify_admin_token)):
    posts = db.query(Post).all()
    return posts

@router.get("/admin/users")
def get_users(db: Session = Depends(get_db), _: dict = Depends(verify_admin_token)):
    users = db.query(User).filter(User.is_admin == False).all()
    return users

@router.get("/admin/comments")
def get_comments(db: Session = Depends(get_db), _: dict = Depends(verify_admin_token)):
    comments = db.query(Comment).all()
    return comments

@router.delete("/admin/{item_type}/{item_id}")
def delete_item(
    item_type: str,
    item_id: int,
    db: Session = Depends(get_db),
    _: dict = Depends(verify_admin_token)
):
    try:
        if item_type == "posts":
            item = db.query(Post).filter(Post.id == item_id).first()
        elif item_type == "users":
            item = db.query(User).filter(User.id == item_id, User.is_admin == False).first()
        elif item_type == "comments":
            item = db.query(Comment).filter(Comment.id == item_id).first()
        else:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Invalid item type: {item_type}"
            )

        if not item:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"{item_type} with id {item_id} not found"
            )

        db.delete(item)
        db.commit()
        return {"message": f"{item_type} deleted successfully"}

    except Exception as e:
        logger.error(f"Error deleting {item_type}: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error deleting {item_type}"
        )

@router.delete("/admin/users/email/{email}")
async def delete_user_by_email(
    email: str,
    db: Session = Depends(get_db),
    _: dict = Depends(verify_admin_token)
):
    """Delete a user by email"""
    try:
        user = db.query(User).filter(User.email == email).first()
        if not user:
            raise HTTPException(status_code=404, detail="User not found")

        db.delete(user)
        db.commit()
        return {"message": "User deleted successfully"}
    except Exception as e:
        db.rollback()
        raise HTTPException(status_code=500, detail=str(e))

File: lost-and-found/src/routes/adminRoutes.js
const express = require('express');
const router = express.Router();
const jwt = require('jsonwebtoken');
const bcrypt = require('bcrypt');
const adminAuth = require('../middleware/adminAuth');
const User = require('../models/User');
const Post = require('../models/Post');
const Comment = require('../models/Comment');

// Admin login
router.post('/login', async (req, res) => {
    try {
        const { username, password } = req.body;
        
        // Check against admin credentials from environment variables
        if (username !== process.env.ADMIN_USERNAME || 
            password !== process.env.ADMIN_PASSWORD) {
            return res.status(401).json({ message: 'Invalid credentials' });
        }

        // Generate admin token
        const token = jwt.sign(
            { isAdmin: true },
            process.env.ADMIN_JWT_SECRET,
            { expiresIn: '24h' }
        );

        res.json({ token });
    } catch (error) {
        res.status(500).json({ message: 'Server error' });
    }
});

// Get all posts
router.get('/posts', adminAuth, async (req, res) => {
    try {
        const posts = await Post.find()
            .populate('author', 'username email')
            .sort({ createdAt: -1 });
        res.json(posts);
    } catch (error) {
        res.status(500).json({ message: 'Error fetching posts' });
    }
});

// Delete a post
router.delete('/posts/:id', adminAuth, async (req, res) => {
    try {
        await Post.findByIdAndDelete(req.params.id);
        // Also delete associated comments
        await Comment.deleteMany({ post: req.params.id });
        res.json({ message: 'Post deleted successfully' });
    } catch (error) {
        res.status(500).json({ message: 'Error deleting post' });
    }
});

// Get all users
router.get('/users', adminAuth, async (req, res) => {
    try {
        const users = await User.find()
            .select('-password')
            .sort({ createdAt: -1 });
        res.json(users);
    } catch (error) {
        res.status(500).json({ message: 'Error fetching users' });
    }
});

// Delete a user
router.delete('/users/:id', adminAuth, async (req, res) => {
    try {
        await User.findByIdAndDelete(req.params.id);
        // Delete all posts and comments by this user
        await Post.deleteMany({ author: req.params.id });
        await Comment.deleteMany({ author: req.params.id });
        res.json({ message: 'User deleted successfully' });
    } catch (error) {
        res.status(500).json({ message: 'Error deleting user' });
    }
});

// Get all comments
router.get('/comments', adminAuth, async (req, res) => {
    try {
        const comments = await Comment.find()
            .populate('author', 'username')
            .populate('post', 'title')
            .sort({ createdAt: -1 });
        res.json(comments);
    } catch (error) {
        res.status(500).json({ message: 'Error fetching comments' });
    }
});

// Delete a comment
router.delete('/comments/:id', adminAuth, async (req, res) => {
    try {
        await Comment.findByIdAndDelete(req.params.id);
        res.json({ message: 'Comment deleted successfully' });
    } catch (error) {
        res.status(500).json({ message: 'Error deleting comment' });
    }
});

module.exports = router;
